### PR TITLE
Add OAuth authentication support

### LINKED ISSUE #142: POC - "Adventurer's Log"  subsection of website
Something like this:

screenie

Basically we will somehow create new subsection of the website (or extract the XP tracker parts etc to subdomain) what will contain XP tracker, tags, loot tracker and overview at home (or login page).

Why subsection? Atm if we would want to put everything into current website navbar, it simply do not fits. Also, it brings the focus on actual important stuff in this subsection (e.g wiki/features/api links and social links are irrelevant here).

We can also extend this later to connect to @Abextm 's proposed local data sharing to grab real time data from the game and display it on website.

### LINKED ISSUE #165: Add support for OAuth authentication
OAuth authentication support is required for future features for website (first what comes to mind is loot tracker).

Implementation should be (from @Adam- 's Discord messages):
* website integration:
runelite.net/login -> google oauth -> runelite.net/oauth -> have the client browser post the oauth token to api.runelite.net to get a runelite session token -> store that token in cookies

* client integration:
i click on a button in the client and it brings me to the site and im automatically authed

### LINKED ISSUE #173: Export Collection log as JSON
Refer to https://github.com/runelite/runelite/issues/7529

TLDR: Allow us to export our entire collection log as a JSON file for external use.

### PR DESCRIPTION
**Pull request by deathbeam**
Resolves #165
Resolves #173 
Closes #142 

Signed-off-by: Tomas Slusny <slusnucky@gmail.com>

### CODE REVIEW

File: src/components/account/GrandExchange.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      <div class="ml-auto">
+        <small class="d-block">{ago(record.date)}</small>
+      </div>
+    </div>
+  </a>
+)
+
+class GrandExchange extends Component {
+  componentDidMount() {
+    this.props.fetchReleases().then(() => this.props.fetchGe())
```
Comment (arthurtyukayev) [inline src/components/account/GrandExchange.js]: This can be rewritten to be a bit cleaner using async/await

```javascript
const fetchInitialData = async () => {
  const { fetchReleases, fetchGe } = this.props;
  await fetchReleases();
  await fetchGe();
}

componentDidMount() => this.fetchInitialData()
```

File: src/components/account/Home.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          </li>
+        ))}
+      </ul>
+    </div>
+  )
+}
+
+class Home extends Component {
+  componentDidMount() {
+    this.props.fetchReleases().then(() => this.props.fetchConfig())
```
Comment (arthurtyukayev) [inline src/components/account/Home.js]: Same as before, I would extract this into another async function and simply call it in `componentDidMount`.

File: src/components/account/LootTracker.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      <h5 class="mb-1">{record.eventId}</h5>
+      <small>{record.type}</small>
+    </div>
+    <div class="mb-1 row">{record.drops.map(buildDrop)}</div>
+  </li>
+)
+
+class LootTracker extends Component {
+  componentDidMount() {
+    this.props.fetchReleases().then(() => this.props.fetchLoot())
```
Comment (arthurtyukayev) [inline src/components/account/LootTracker.js]: See previous.

File: src/modules/account.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+        ws.onclose = msg => reject(msg)
+        ws.onerror = msg => reject(msg)
+      })
+
+      return await sessionPromise
+    },
+    LOGOUT: () => async (dispatch, getState) => {
+      const version = getLatestRelease(getState()).name
+      const localUuid = getState().account.uuid
```
Comment (arthurtyukayev) [inline src/modules/account.js]: I would refactor using destructing, before this will crash if account is undefined or null.

```javascript
const state = getState();
const { account = {} } = state;
const version = getLatestVersion(state).name
const localUuid = account.uuid // This way app wouldn't crash, uuid would just `undefined`
```

File: src/modules/account.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            'RUNELITE-AUTH': localUuid
+          }
+        })
+      } finally {
+        dispatch(resetSession())
+      }
+    },
+    SESSION_CHECK: () => async (dispatch, getState) => {
+      const version = getLatestRelease(getState()).name
+      const localUuid = getState().account.uuid
```
Comment (arthurtyukayev) [inline src/modules/account.js]: Same as above. This will crash.

File: src/modules/account.js (hunk 4 above)
Comment (arthurtyukayev) [reply src/modules/account.js]: Make sure to check `localUuid`.

File: src/modules/config.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  /^(killcount)\.(.+)\.([^.]+)$/,
+  /^([^.]+)\.(.+)$/
+]
+
+// Actions
+export const { fetchConfig, setConfig, changeAccount } = createActions(
+  {
+    FETCH_CONFIG: () => async (dispatch, getState) => {
+      const version = getLatestRelease(getState()).name
+      const uuid = getState().account.uuid
```
Comment (arthurtyukayev) [inline src/modules/config.js]: Same as before. This can crash.

File: src/modules/account.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+          ws.close()
+        }
+
+        ws.onclose = msg => reject(msg)
+        ws.onerror = msg => reject(msg)
+      })
+
+      return await sessionPromise
+    },
+    LOGOUT: () => async (dispatch, getState) => {
```
Comment (arthurtyukayev) [inline src/modules/account.js]: Missing `try/catch` here.
Using `async/await` with fetch/XHR without error handling will crash the app.

File: src/modules/config.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  /^(timetracking)\.(.+)\.([0-9]+\.[0-9]+)$/,
+  /^(timetracking)\.(.+)\.(autoweed)$/,
+  /^(killcount)\.(.+)\.([^.]+)$/,
+  /^([^.]+)\.(.+)$/
+]
+
+// Actions
+export const { fetchConfig, setConfig, changeAccount } = createActions(
+  {
+    FETCH_CONFIG: () => async (dispatch, getState) => {
```
Comment (arthurtyukayev) [inline src/modules/config.js]: Same as before, need error handling

File: src/modules/ge.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import api from '../api'
+import { getLatestRelease } from './git'
+
+const runeliteApi = api('https://api.runelite.net/')
+const runeliteStaticApi = api('https://static.runelite.net/')
+
+// Actions
+export const { fetchGe, setGe, setGeRange } = createActions(
+  {
+    FETCH_GE: () => async (dispatch, getState) => {
```
Comment (arthurtyukayev) [inline src/modules/ge.js]: Same as before, needs error handling.

File: src/modules/ge.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+      const version = getLatestRelease(getState()).name
+      const uuid = getState().account.uuid
```
Comment (arthurtyukayev) [inline src/modules/ge.js]: Same as before, could use safer destructuring.

File: src/modules/loot.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import { createActions, handleActions } from 'redux-actions'
+import api from '../api'
+import { getLatestRelease } from './git'
+
+const runeliteApi = api('https://api.runelite.net/')
+
+// Actions
+export const { fetchLoot, setLoot, setLootRange } = createActions(
+  {
+    FETCH_LOOT: () => async (dispatch, getState) => {
```
Comment (arthurtyukayev) [inline src/modules/loot.js]: Same as before.

File: src/modules/account.js (hunk 4 above)
Comment (deathbeam) [reply src/modules/account.js]: account will never be undefined because its part of initial redux state and so is always created before everything else.

File: src/modules/account.js (hunk 7 above)
Comment (deathbeam) [reply src/modules/account.js]: try/catch is already in thunk-middleware

**Comment by deathbeam** [comment]
@arthurtyukayev The async/await improvement for initial fetch for components applies to rest of the site too, so I guess opening separate enhancement issue for that would be better so it can be done at once

**Comment by arthurtyukayev** [comment]
> @arthurtyukayev The async/await improvement for initial fetch for components applies to rest of the site too, so I guess opening separate enhancement issue for that would be better so it can be done at once

Sure thing, and error-handling in the thunk-middleware is a good idea, I didn't actually realize that, really nice.

**Comment by arthurtyukayev** [comment]
@deathbeam The login button blends into the background, maybe switch to non-outline version?

image

**Comment by arthurtyukayev** [comment]
@deathbeam Overall the code looks good. The only thing that I am seeing is that you're not using loading states for any of the API requests. I would add them for all of the new API stores. GE, Loot Tracker, and Farming Tracker.

**Comment by deathbeam** [comment]
I will update the login button. For loading states, thunk-middleware is handling that, sort of, its displaying loading bar on top

**Comment by deathbeam** [comment]
Depends on https://github.com/runelite/runelite/pull/7843 to add sorting to loot tracker (atm it is mess).
